### ISSUE TITLE
Use of Sharerootfs leads to "missing filesystems" (when "drive letters" get shifted afterwards)

### ISSUE THREAD

**Issue by momo42** [openmediavault/openmediavault#1212]
**Describe the bug**

In my hardware setup I have four drive bays populated with data harddisks and one ssd containing the os. While installing OMV I remove the data disks to prevent any loss of data. So at this point of time the system ssd is /dev/sda.
After initial setup I install and configure some addons needed for later operation - also sharerootfs (6.0-3)!

This is how the drives tab looks like at this point:
![Screenshot 2022-02-09 at 07-09-57 openmediavault Workbench](https://user-images.githubusercontent.com/28578876/153729262-44d1cf25-4c00-4039-aff3-30eb7f088c72.png)

Sharerootfs makes the system partition (/dev/sda1 at this point) showing up on the file systems tab. Everything is fine.
Then I attach the data drives and the mess begins.
![Screenshot 2022-02-09 at 03-45-28 openmediavault Workbench](https://user-images.githubusercontent.com/28578876/153729295-d39d28df-6a3d-4906-bab1-7edaeb58edde.png)

For some reasons I cannot explain the data disks are always put in front and the system disk is shifted to the end. In this case with only one harddrive the system ssd becomes /dev/sdb.
As a result I get this error message in the file systems tab:
![Screenshot 2022-02-09 at 03-46-11 openmediavault Workbench](https://user-images.githubusercontent.com/28578876/153729320-6d0bfc39-9a53-48ed-a369-df2af22dc37b.png)

At the dashboad the system file system is shown as "missing":
![Screenshot 2022-02-09 at 03-48-19 openmediavault Workbench](https://user-images.githubusercontent.com/28578876/153729326-37824794-7061-42d1-abcb-3f1abf405fab.png)

and also disappears from the dashboard file system list:
![Screenshot 2022-02-09 at 03-48-40 openmediavault Workbench](https://user-images.githubusercontent.com/28578876/153729331-b0c9625a-7052-4def-83f4-2dbe8c62329b.png)

The corresponding forum thread is here:
https://forum.openmediavault.org/index.php?thread/42041-use-of-sharerootfs-leads-to-missing-filesystems/

Two years ago I already faced a similar problem in OMV5
https://forum.openmediavault.org/index.php?thread/31291-omv-remembers-drive-letters-during-installation-and-creates-ghost-file-systems-a/, but never found the reason nor solution to that.

Back then I had no idea that sharerootfs might be responsible for those "missing" ghost entries in the OMV5 file systems tab but I'm confident now that sharerootfs in combination with my system shifting the "drive letters" when attaching more drives is the problem.

When I realized that I uninstalled and reinstalled sharerootfs and everything went back to normal!

I have no idea what sharerootfs does exactly (and WHEN it does it) but it is obviously overwhelmed when a newly attached drive pushes the system drive back in the sequence.

**To Reproduce**
Reproducing this behavior in Virtualbox is a bit hard but I managed to do so.
1. Install OMV on a single virtual hard drive attached to SATA port 0 (default).
2. In OMV run updates, install sharerootfs and shut down the vm.
3. Add another virtual hard drive to the vm and reorder them (in virtualbox) so that the system disk is on SATA port 1 and the new one is on SATA port 0.
4. When you try to start the vm, you now get an error message (no bootable medium found).
5. So press F12 immediately after starting the vm and type "2" to select the right virtual hard disk for startup.

After that you'll see the exact same problem described.

I'm not a dev and I didn't look into the details of sharerootfs but I fear its magic has to be repeated every time the drive order changes to prevent this bug. :/

Side note: On my server setup the system ssd is set to first place in bios boot order. But nevertheless OMV pushes it behind the data disks at system start - even some usb disks get lower letters than my system disk sometimes if present at system start. When I attach all 4 data drives, the system ssd ends up as being /dev/sde. And every time such a shift happens these errors have to be repaired with uninstalling/installing sharerootfs again.

**Expected behavior**
sharerootfs should be able to deal with the drive order being shifted after installation.

**openmediavault Server (please complete the following information):**

- OS version: 5.14.0-0.bpo.2-amd64 # 1 SMP Debian 5.14.9-2~bpo11+1 (2021-10-10) x86_64 GNU/Linux
- openmediavault version: 6.0.11-1

**Additional context**

SMART dashboard widget seems to be a bit confused too by the letter shifting.
![Screenshot 2022-02-12 at 22-57-19 openmediavault Workbench](https://user-images.githubusercontent.com/28578876/153729796-610cf7ec-d2e7-4802-9549-9603dc2d128d.png)
(Screenshot shows recent setup with system ssd, two data hdds and one usb drive (with no SMART available) being attached. I'd expect these buttons being ordered alphabetically but instead they seem to be in chronological(?) order.)

**Comment by votdev** [comment]
Could you please run the command `omv-confdbadm read --prettify conf.system.filesystem.mountpoint` and post the output here?

Is it possible that you execute the command `salt-call --out=json mount.fstab | jq --compact-output '.[] | .["/"]'
{"device":"UUID=d614af06-856c-4d2e-ac62-bc21d6852507","fstype":"ext4","opts":["rw","discard","errors=remount-ro"],"dump":"0","pass":"1"}
` after the initial setup to get the origin data that is used to configure the mount point in OMV?

**Comment by votdev** [comment]
> Side note: On my server setup the system ssd is set to first place in bios boot order. But nevertheless OMV pushes it behind the data disks at system start - even some usb disks get lower letters than my system disk sometimes if present at system start. When I attach all 4 data drives, the system ssd ends up as being /dev/sde. And every time such a shift happens these errors have to be repaired with uninstalling/installing sharerootfs again.

This is how the kernel assigns device names, first comes first serve. Because of that device names like `/dev/sda` should be prevented wherever possible. This is the default behaviour of OMV. OMV tries to use predictable device names, e.g. `/dev/disk/by-uuid/xxxx` or `/dev/disk/by-path/pci-0000:00:03.0-scsi-0:0:1:0`, if possible because those are, as the name indicates, predictable and will not change by random.

**Comment by momo42** [comment]
> Could you please run the command `omv-confdbadm read --prettify conf.system.filesystem.mountpoint` and post the output here?
> 
> Is it possible that you execute the command `salt-call --out=json mount.fstab | jq --compact-output '.[] | .["/"]'` after the initial setup to get the origin data that is used to configure the mount point in OMV? I know the setup has to be done from scratch again to get this data.

I promised my wife not to mess around with the real hardware anymore :D but I'll setup a new VM for testing that. Hope this is ok. Will report back my in the next days.

**Comment by momo42** [comment]
> > Side note: On my server setup the system ssd is set to first place in bios boot order. But nevertheless OMV pushes it behind the data disks at system start - even some usb disks get lower letters than my system disk sometimes if present at system start. When I attach all 4 data drives, the system ssd ends up as being /dev/sde. And every time such a shift happens these errors have to be repaired with uninstalling/installing sharerootfs again.
> 
> This is how the Linux kernel assigns device names, first comes first serve. Because of that device names like `/dev/sda` should be prevented wherever possible. This is the default behaviour of OMV. OMV tries to use predictable device names, e.g. `/dev/disk/by-uuid/xxxx` or `/dev/disk/by-path/pci-0000:00:03.0-scsi-0:0:1:0`, if possible because those are, as the name indicates, predictable and will not change by random.

Unlike other people who complain about it on the forum, I' m generally fine with the linux kernels behavior and I like the way OMV handles it. But this server (HP Proliant Microserver Gen8 with system ssd on PCIE card) is indeed the first hardware I see, where the first device in boot order isn't also /dev/sda. In fact most of the times it is the last one here. Even USB drives, that are present at boot time get lower device names.

**Comment by votdev** [comment]
I think the issue will be fixed by the fix for https://github.com/openmediavault/openmediavault/issues/1215.

**Comment by momo42** [comment]
@votdev Please excuse my late reply. Some COVID cases in my family kept me very busy for the last two months.

> I think the issue will be fixed by the fix for #1215.

Unfortunately not. I set up a testing vm yesterday and can report back to you that the behavior didn't change at all with all updates installed (System info says OMV version 6.0.21-1).

I executed your commands to track the issue:

This commands' output never changes:
```
root@omv6-test:~# salt-call --out=json mount.fstab | jq --compact-output '.[] | .["/"]'
{"device":"UUID=1aad5bed-ee92-404d-84c1-2680a45664a8","fstype":"ext4","opts":["errors=remount-ro"],"dump":"0","pass":"1"}
```

On a fresh installed system (without sharerootfs installed) this command returns just empty brackets (as expected, I assume).
```
root@omv6-test:~# omv-confdbadm read --prettify conf.system.filesystem.mountpoint
[]
```
after installing sharerootfs it looks like this:
```
root@omv6-test:~# omv-confdbadm read --prettify conf.system.filesystem.mountpoint
[
    {
        "comment": "",
        "dir": "/",
        "freq": 0,
        "fsname": "/dev/sda1",
        "hidden": true,
        "opts": "errors=remount-ro",
        "passno": 1,
        "type": "ext4",
        "usagewarnthreshold": 0,
        "uuid": "79684322-3eac-11ea-a974-63a080abab18"
    }
]
```
When I attach a second virtual disk and change the drive order in Virtualbox so that the system drive gets /dev/sdb nothing happens here. Still the same output.
Uninstalling and reinstalling sharerootfs fixes the display bug as before and leads to that output:
root@omv6-test:~# omv-confdbadm read --prettify conf.system.filesystem.mountpoint
```
[
    {
        "comment": "",
        "dir": "/",
        "freq": 0,
        "fsname": "/dev/sdb1",
        "hidden": true,
        "opts": "errors=remount-ro",
        "passno": 1,
        "type": "ext4",
        "usagewarnthreshold": 0,
        "uuid": "79684322-3eac-11ea-a974-63a080abab18"
    }
]
```

Here some screenshots with boot drive on first port, data drive on second port:
![Screenshot 2022-04-12 at 07-02-44 openmediavault Workbench](https://user-images.githubusercontent.com/28578876/162887167-7743f74d-a5a1-4c05-a6d2-0bd54d4b1381.png)
![Screenshot 2022-04-12 at 07-04-32 openmediavault Workbench](https://user-images.githubusercontent.com/28578876/162887175-d6ec2d5d-7f93-4297-94dc-2548659a9e17.png)

This is how it looks like when drive order changed and system drive was pushed to second place (error case):
![Screenshot 2022-04-12 at 07-05-56 openmediavault Workbench](https://user-images.githubusercontent.com/28578876/162887576-febc1d87-64d0-4101-b40e-695c337b7d8e.png)
![Screenshot 2022-04-12 at 07-06-07 openmediavault Workbench](https://user-images.githubusercontent.com/28578876/162887584-1fc2ec1f-f5d0-4957-9a68-745d36a14d17.png)

After uninstalling and reinstalling sharerootfs things are fixed:
![Screenshot 2022-04-12 at 07-09-09 openmediavault Workbench](https://user-images.githubusercontent.com/28578876/162887653-c8428c17-84d6-496c-8eec-7058331a35b9.png)
![Screenshot 2022-04-12 at 07-09-19 openmediavault Workbench](https://user-images.githubusercontent.com/28578876/162887666-9ca4cb27-f971-47b8-9a36-7c28dbe154a5.png)

Next change of drive order breaks it again ...
